### DISCUSSION TITLE
Creating a hierarchy of prebuilds

### DISCUSSION THREAD

**Discussion by seesharprun** [community/community#65602] [Codespaces] [answered]
### Select Topic Area

Question

### Body

I have a repository with a primary `devcontainer.json` file and multiple "alternate" profiles. I have prebuilds set for all of the profiles.

This is the primary file:

```json
{
    "name": "All patterns",
    "image": "mcr.microsoft.com/devcontainers/dotnet:6.0",
    "features": {
        "ghcr.io/devcontainers/features/azure-cli:1": {
            "installBicep": true,
            "extensions": "cosmosdb-preview"
        }
    },
    "containerEnv": {
        "ASPNETCORE_ENVIRONMENT": "Development"
    },
    "customizations": {
        "vscode": {
            "extensions": [
                "ms-dotnettools.csharp",
                "ms-azuretools.vscode-bicep"
            ]
        }
    }
}
```

I'm following the monorepo pattern where we have multiple example patterns in a single repo. The only difference between this primary file and the alternates is the workspace mount. Each alternate mounts to a subfolder and runs a build script so you can focus on a single pattern. Here's an example:

```json
{
    "name": "Preallocation pattern",
    "image": "mcr.microsoft.com/devcontainers/dotnet:6.0",
    "workspaceMount": "source=${localWorkspaceFolder},target=/workspace,type=bind",
    "workspaceFolder": "/workspace/preallocation/source/",
    "updateContentCommand": "dotnet build Cosmos_Patterns_Preallocation.csproj",
    "features": {
        "ghcr.io/devcontainers/features/azure-cli:1": {
            "installBicep": true,
            "extensions": "cosmosdb-preview"
        }
    },
    "containerEnv": {
        "ASPNETCORE_ENVIRONMENT": "Development"
    },
    "customizations": {
        "codespaces": {
            "openFiles": [
                "readme.md"
            ]
        },
        "vscode": {
            "settings": {
                "git.openRepositoryInParentFolders": "always"
            },
            "extensions": [
                "ms-dotnettools.csharp",
                "ms-azuretools.vscode-bicep"
            ]
        }
    }
}
```

Overall, it technically works. Unfortunately, I have almost ten patterns running prebuilds. This means that we must rebuild the same image that is almost identical over and over again saturating the GitHub runners. **Is there a more efficient way to do this**?

Is there a concept of inheritance with devcontainers and prebuilds? Can I use a GitHub action to prebuild most of the image and just  apply the mount per profile? Has anyone tried something similar? My "scooby sense" says this is not the best way to do it, but I can't find documentation on a better alternative.

**Comment by arafajs** [comment, marked as answer]
Yes, you can definitely use inheritance and reuse with dev containers to avoid rebuilding the same image multiple times. Here are a couple approaches I've seen used for this: **GitHub Actions to build a base image**

1. Have a GitHub workflow that builds a general purpose "base" dev container image with all the common tools, runtimes, etc.installed.
2. Push this base image to a registry like GitHub Container Registry.
3. Then in each devcontainer profile, use "image": "ghcr.io/myorg/mybaseimage:latest" to inherit from that base image.
4. Use lifecycle hooks like postCreateCommand to customize each profile.

**Multiple devcontainer.json files**

- Put the common aspects into a shared base.devcontainer.json.
- In each profile, create a devcontainer.json that extends the base:

```
{
  "extends": "./base.devcontainer.json",
  
  "customizations": {
    // profile specific customizations
  }
}
```

**Reply by seesharprun** [reply]
Thank you so much! I will try this out right away and come back and mark your response as an answer if I can get this to work.
